### PR TITLE
Hwp metadata update

### PR DESCRIPTION
**Pull request by ykyohei**
This PR updates hwp_angles metadata and implement apply_hwp_angle_model

- Add draft of apply_hwp_angle_model https://github.com/simonsobs/sotodlib/discussions/735
  - We still need to implemenet hwp_angle_model metadata
  - Previously make_hwp_solution was correcting relative hwp rotation direction between observations using configured method, but removed this correction from make_hwp_solution to implement more robust correction in apply_hwp_angle_model
- Add sotodlib version to hwp_angles metadata https://github.com/simonsobs/sotodlib/issues/776
- Split hwp_angles metadata more https://simonsobs.slack.com/archives/C05K7CSFTPX/p1712183463493169
  - This is necessary for more efficient hwp metadata data transfer
  - Save fast/slow data into different HDF5 files.
  - Split HDF5 file by first 4 digits of unixtime -> 5 digits for unixtime.

### CODE REVIEW

**Comment by 17-sugiyama** [comment]
It seems good for me!
In [this line](https://github.com/simonsobs/sotodlib/blob/eab5872ee813393f3368c5f9e2052cae11c72111/sotodlib/hwp/hwp_angle_model.py#L62), the different angle offset is defined for two bands. I think we can remove the band dependency if we don't apply it.

**Review by 17-sugiyama — APPROVED** [review]

**Comment by ykyohei** [comment]
I make this PR ready for review

- The new hwp metadata is available here
/so/home/ykyohei/metadata/satp3/manifests/hwp_angles/satp3_hwp_angles_240430m
- apply_hwp_angle_model return single approximate hwp_angle without passband dependence
- Junna checked the consistency between multiple rotation direction estimations and the benefit of using multiple estimations
https://simonsobs.atlassian.net/wiki/spaces/PRO/pages/257392641/Daily+analysis#2024%2F4%2F26
- This PR should be merged with https://github.com/simonsobs/site-pipeline-configs/pull/67

File: sotodlib/hwp/g3thwp.py [hunk 1]
```diff
@@ -905,87 +958,35 @@ def write_solution_h5(self, tod, output=None, h5_address=None, load_h5=True):
...
+            if not 'counter' + suffix in self.data.keys():
                 logger.warning('No HWP data in the specified timestamps.')
                 self._write_solution_h5_logger = 'No HWP data'
                 continue
 
             # version 1
             # calculate HWP angle
             try:
-                solved = solved | self.analyze(data, mod2pi=False, suffix=suffix)
+                solved = solved | self.analyze(self.data, mod2pi=False, suffix=suffix)
```
Comment (mhasself) [inline sotodlib/hwp/g3thwp.py:989]: This is a very strange line... the operator `|` isn't defined for dicts.  Am I missing something?

Comment (mhasself) [reply sotodlib/hwp/g3thwp.py:989]: You probably mean "or", right?  Has this code been tested?

Comment (ykyohei) [reply sotodlib/hwp/g3thwp.py:989]: This is merging two dictionaries for two encoders. This way is supported from python 3.9.0
https://stackoverflow.com/questions/38987/how-do-i-merge-two-dictionaries-in-a-single-expression-in-python

Comment (ykyohei) [reply sotodlib/hwp/g3thwp.py:989]: And yes, this is tested. We have minor modification in this PR, but the this code (operator `|`) has been used from previous version of metadata 240301m.

Comment (mhasself) [reply sotodlib/hwp/g3thwp.py:989]: Ah, I have not seen that... and was checking on python 3.8.

We are allowed to require py3.9 now but I will note you could have simply done
```
solved.update(self.analyze(self.data, mod2pi=False, suffix=suffix))
```
... and that would have been python2 compatible!

Thanks.

**Comment by ykyohei** [comment]
@mhasself Please let me confirm. Is it okay to moving on merging this and updating the site-pipeline with Yilun? 
I guess we need a review for apply_hwp_angle_model.py.

**Review by mhasself — CHANGES_REQUESTED** [review]
Thanks for all the work on this and sorry for the delay in responding.

See a few comments inline.  After those, I think we can merge this.

However, I want to point out that the API for the function will likely change a bit, in the near future.  `apply_hwp_angle_model(tod)` will be fine, and fairly portable... but the other arguments will change.  And the parameters themselves, which are currently hard-coded, will become metadata products.  But that can wait.

File: sotodlib/hwp/hwp_angle_model.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        'satp3': -2.29,
+    }
+
+    offset = encoder_offset + \
+        encoder_assembly_offset[telescope][hwp.primary_encoder] + \
+        ahwp_offset[telescope]
+
+    # apply correction
+    tod.hwp_angle = np.mod(
+        sign*tod.hwp_angle + np.deg2rad(offset), 2*np.pi)
```
[the remark this answers is not in the document]
Comment (ykyohei) [reply sotodlib/hwp/hwp_angle_model.py]: I found `accept_new` and `accept_merge` option of aman.wrap are not doing anything, so checked it by _assignments.
https://github.com/simonsobs/sotodlib/blob/12b4b91265269bb38d1272c44e86d341a322806f/sotodlib/core/axisman.py#L555-L556

File: sotodlib/hwp/hwp_angle_model.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        'satp3': {'pid': -1, 'offcenter':  1, },
+    }
+
+    # check available sign estimation methods
+    available_signs = []
+
+    methods = ['pid', 'offcenter']
+    for method in methods:
+        _sign = getattr(hwp, method + '_direction')
+        if _sign != 0:
```
[the remark this answers is not in the document]
Comment (ykyohei) [reply sotodlib/hwp/hwp_angle_model.py]: None does not happen.

**Comment by ykyohei** [comment]
@mhasself Addressed your suggestions and tested that `apply_hwp_angle_model` works with new metadata.

**Review by mhasself — APPROVED** [review]
Thanks for the adjustments!

**Pull request merged**